README.rst
=======================================
Welcome to python-binance-chain v0.1.20
=======================================

.. image:: https://img.shields.io/pypi/v/python-binance-chain.svg
    :target: https://pypi.python.org/pypi/python-binance-chain

.. image:: https://img.shields.io/pypi/l/python-binance-chain.svg
    :target: https://pypi.python.org/pypi/python-binance-chain

.. image:: https://img.shields.io/travis/sammchardy/python-binance-chain.svg
    :target: https://travis-ci.org/sammchardy/python-binance-chain

.. image:: https://img.shields.io/coveralls/sammchardy/python-binance-chain.svg
    :target: https://coveralls.io/github/sammchardy/python-binance-chain

.. image:: https://img.shields.io/pypi/wheel/python-binance-chain.svg
    :target: https://pypi.python.org/pypi/python-binance-chain

.. image:: https://img.shields.io/pypi/pyversions/python-binance-chain.svg
    :target: https://pypi.python.org/pypi/python-binance-chain

This is an unofficial Python3 wrapper for the `Binance Chain API <https://binance-chain.github.io/api-reference/dex-api/paths.html>`_. I am in no way affiliated with Binance, use at your own risk.


PyPi
  https://pypi.python.org/pypi/python-binance-chain

Source code
  https://github.com/sammchardy/python-binance-chain


Features
--------

- Support for Testnet and Production `environments <#environments>`_, along with user defined environment
- HTTP API `sync <#quick-start>`_ and `async <#async-http-client>`_ implementations
- `Async Websockets <#websockets>`_ with auto-reconnection and backoff retry algorithm
- HTTP RPC Node `sync <#node-rpc-http>`_ and `async <#node-rpc-http-async>`_ implementations
- Advanced async `Pooled HTTP RPC Node client <#pooled-node-rpc-client>`_ spreading requests over available peers
- `Async Node RPC Websockets <#node-rpc-websockets>`_ with auto-reconnection and backoff retry algorithm
- `Wallet <#wallet>`_ creation from private key or mnemonic or new wallet with random mnemonic
- Wallet handling account sequence for transactions
- Broadcast Transactions over `HTTP <#broadcast-messages-on-httpapiclient>`_ and `RPC <#node-rpc-http>`_ with helper classes for limit buy and sell
- `Sign transactions <#sign-transaction>`_ and use the signed message how you want
- `Ledger hardware wallet <#ledger>`_ device (Ledger Blue, Nano S & Nano X) support for signing messages
- Async `Depth Cache <#depth-cache>`_ to keep a copy of the order book locally
- `Signing Service Support <#signing-service>`_ for `binance-chain-signing-service <https://github.com/sammchardy/binance-chain-signing-service>`_
- Support for HTTP and HTTPS proxies and to override `Requests and AioHTTP settings <requests-and-aiohttp-settings>`_
- `UltraJson <https://github.com/esnme/ultrajson>`_ the ultra fast JSON parsing library for efficient message handling
- Strong Python3 typing to reduce errors
- pytest `test suite <#running-tests>`_
- Response exception handling

Read the `Changelog <https://python-binance-chain.readthedocs.io/en/latest/changelog.html>`_


Recommended Resources
---------------------

- `Binance Chain Forum <https://community.binance.org/>`_
- `Binance Chain Telegram <https://t.me/BinanceDEXchange>`_
- `Binance Chain API <https://binance-chain.github.io/>`_
- `Tendermint Docs <https://tendermint.com/docs/>`_
- `Get Testnet Funds <https://www.binance.vision/tutorials/binance-dex-funding-your-testnet-account>`_


Quick Start
-----------

.. code:: bash

    pip install python-binance-chain

If having issues with secp256k1 check the `Installation instructions for the sec256k1-py library <https://github.com/ludbb/secp256k1-py#installation>`_

If using the production server there is no need to pass the environment variable.

.. code:: python

    from binance_chain.http import HttpApiClient
    from binance_chain.constants import KlineInterval
    from binance_chain.environment import BinanceEnvironment

    # initialise with Testnet environment
    testnet_env = BinanceEnvironment.get_testnet_env()
    client = HttpApiClient(env=testnet_env)

    # Alternatively pass no env to get production
    prod_client = HttpApiClient()

    # connect client to different URL
    client = HttpApiClient(api_url='https://yournet.com')

    # get node time
    time = client.get_time()

    # get node info
    node_info = client.get_node_info()

    # get validators
    validators = client.get_validators()

    # get peers
    peers = client.get_peers()

    # get account
    account = client.get_account('tbnb185tqzq3j6y7yep85lncaz9qeectjxqe5054cgn')

    # get account sequence
    account_seq = client.get_account_sequence('tbnb185tqzq3j6y7yep85lncaz9qeectjxqe5054cgn')

    # get markets
    markets = client.get_markets()

    # get fees
    fees = client.get_fees()

    # get order book
    order_book = client.get_order_book('NNB-0AD_BNB')

    # get klines
    klines = client.get_klines('NNB-338_BNB', KlineInterval.ONE_DAY)

    # get closed orders
    closed_orders = client.get_closed_orders('tbnb185tqzq3j6y7yep85lncaz9qeectjxqe5054cgn')

    # get open orders
    open_orders = client.get_open_orders('tbnb185tqzq3j6y7yep85lncaz9qeectjxqe5054cgn')

    # get open orders
    ticker = client.get_ticker('NNB-0AD_BNB')

    # get open orders
    trades = client.get_trades(limit=2)

    # get open orders
    order = client.get_order('9D0537108883C68B8F43811B780327CE97D8E01D-2')

    # get open orders
    trades = client.get_trades()

    # get transactions
    transactions = client.get_transactions(address='tbnb1n5znwyygs0rghr6rsydhsqe8e6ta3cqatucsqp')

    # get transaction
    transaction = client.get_transaction('95DD6921370D74D0459590268B439F3DD49F6B1D090121AFE4B2183C040236F3')

See `API <https://python-binance-chain.readthedocs.io/en/latest/binance-chain.html#module-binance_chain>`_ docs for more information.

Async HTTP Client
-----------------

An implementation of the HTTP Client above using aiohttp instead of requests

Use the async `create` classmethod to initialise an instance of the class.

All methods are otherwise the same as the HttpApiClient


.. code:: python

    from binance_chain.http import AsyncHttpApiClient
    from binance_chain.environment import BinanceEnvironment

    loop = None

    async def main():
        global loop

        env = BinanceEnvironment.get_testnet_env()

        # initialise the class using the classmethod
        client = await AsyncHttpApiClient.create(env)
        wallet = Wallet(private_key=priv_key, env=env)

        print(json.dumps(await client.get_time(), indent=2))

        while True:
            print("doing a sleep")
            await asyncio.sleep(20, loop=loop)


    if __name__ == "__main__":

        loop = asyncio.get_event_loop()
        loop.run_until_complete(main())


Environments
------------

Binance Chain offers a Production system and Testnet.

If using the Production system there is no need to pass an environment as this is the default.

To create and use the Testnet environment is as easy as

.. code:: python

    from binance_chain.environment import BinanceEnvironment

    # initialise with Testnet environment
    testnet_env = BinanceEnvironment.get_testnet_env()

You may also create your own custom environments, this may be useful such as connecting to a Node RPC client

.. code:: python

    from binance_chain.environment import BinanceEnvironment

    # create custom environment
    my_env = BinanceEnvironment(api_url="<api_url>", wss_url="<wss_url>", hrp="<hrp>")


See `API <https://python-binance-chain.readthedocs.io/en/latest/binance-chain.html#module-binance_chain.environment>`_ docs for more information.

Wallet
------

See `API <https://python-binance-chain.readthedocs.io/en/latest/binance-chain.html#module-binance_chain.wallet>`_ docs for more information.

The wallet is required if you want to place orders, transfer funds or freeze and unfreeze tokens.

You may also use the `Ledger Wallet class <#ledger>`_ to utilise your Ledger Hardware Wallet for signing.

It can be initialised with your private key or your mnemonic phrase.

Note that the BinanceEnvironment used for the wallet must match that of the HttpApiClient, testnet addresses will not
work on the production system.

The Wallet class can also create a new account for you by calling the `Wallet.create_random_wallet()` function,
see examples below


**Initialise from Private Key**

.. code:: python

    from binance_chain.wallet import Wallet
    from binance_chain.environment import BinanceEnvironment

    testnet_env = BinanceEnvironment.get_testnet_env()
    wallet = Wallet('private_key_string', env=testnet_env)
    print(wallet.address)
    print(wallet.private_key)
    print(wallet.public_key_hex)

**Initialise from Mnemonic**

.. code:: python

    from binance_chain.wallet import Wallet
    from binance_chain.environment import BinanceEnvironment

    testnet_env = BinanceEnvironment.get_testnet_env()
    wallet = Wallet.create_wallet_from_mnemonic('mnemonic word string', env=testnet_env)
    print(wallet.address)
    print(wallet.private_key)
    print(wallet.public_key_hex)

**Initialise by generating a random Mneomonic**

.. code:: python

    from binance_chain.wallet import Wallet
    from binance_chain.environment import BinanceEnvironment

    testnet_env = BinanceEnvironment.get_testnet_env(, env=testnet_env)
    wallet = Wallet.create_random_wallet(env=env)
    print(wallet.address)
    print(wallet.private_key)
    print(wallet.public_key_hex)

Broadcast Messages on HttpApiClient
-----------------------------------

See `API <https://python-binance-chain.readthedocs.io/en/latest/binance-chain.html#module-binance_chain.messages>`_ docs for more information.

Requires a Wallet to have been created.

The Wallet will increment the request sequence when broadcasting messages through the HttpApiClient.

If the sequence gets out of sync call `wallet.reload_account_sequence(client)`, where client is an instance of HttpApiClient.

**Place Order**

General case

.. code:: python

    from binance_chain.http import HttpApiClient
    from binance_chain.messages import NewOrderMsg
    from binance_chain.wallet import Wallet

    wallet = Wallet('private_key_string')
    client = HttpApiClient()

    # construct the message
    new_order_msg = NewOrderMsg(
        wallet=wallet,
        symbol="ANN-457_BNB",
        time_in_force=TimeInForce.GTE,
        order_type=OrderType.LIMIT,
        side=OrderSide.BUY,
        price=Decimal(0.000396000),
        quantity=Decimal(12)
    )
    # then broadcast it
    res = client.broadcast_msg(new_order_msg, sync=True)

**Limit Order Buy**

.. code:: python

    from binance_chain.messages import LimitOrderBuyMsg

    limit_order_msg = LimitOrderBuyMsg(
        wallet=wallet,
        symbol='ANN-457_BNB',
        price=0.000396000,
        quantity=12
    )

**Limit Order Sell**

.. code:: python

    from binance_chain.messages import LimitOrderSellMsg

    limit_order_msg = LimitOrderSellMsg(
        wallet=wallet,
        symbol='ANN-457_BNB',
        price=0.000396000,
        quantity=12
    )

**Cancel Order**

.. code:: python

    from binance_chain.http import HttpApiClient
    from binance_chain.messages import CancelOrderMsg
    from binance_chain.wallet import Wallet

    wallet = Wallet('private_key_string')
    client = HttpApiClient()

    # construct the message
    cancel_order_msg = CancelOrderMsg(
        wallet=wallet,
        order_id="order_id_string",
        symbol='ANN-457_BNB',
    )
    # then broadcast it
    res = client.broadcast_msg(cancel_order_msg, sync=True)


**Freeze Tokens**

.. code:: python

    from binance_chain.http import HttpApiClient
    from binance_chain.messages import FreezeMsg
    from binance_chain.wallet import Wallet

    wallet = Wallet('private_key_string')
    client = HttpApiClient()

    # construct the message
    freeze_msg = FreezeMsg(
        wallet=wallet,
        symbol='BNB',
        amount=Decimal(10)
    )
    # then broadcast it
    res = client.broadcast_msg(freeze_msg, sync=True)


**Unfreeze Tokens**

.. code:: python

    from binance_chain.http import HttpApiClient
    from binance_chain.messages import UnFreezeMsg
    from binance_chain.wallet import Wallet

    wallet = Wallet('private_key_string')
    client = HttpApiClient()

    # construct the message
    unfreeze_msg = UnFreezeMsg(
        wallet=wallet,
        symbol='BNB',
        amount=Decimal(10)
    )
    # then broadcast it
    res = client.broadcast_msg(unfreeze_msg, sync=True)


**Transfer Tokens**

.. code:: python

    from binance_chain.http import HttpApiClient
    from binance_chain.messages import TransferMsg
    from binance_chain.wallet import Wallet

    wallet = Wallet('private_key_string')
    client = HttpApiClient()

    transfer_msg = TransferMsg(
        wallet=wallet,
        symbol='BNB',
        amount=1,
        to_address='<to address>',
        memo='Thanks for the beer'
    )
    res = client.broadcast_msg(transfer_msg, sync=True)

**Transfer Multiple Tokens**

.. code:: python

    from binance_chain.http import HttpApiClient
    from binance_chain.messages import TransferMsg, Transfer
    from binance_chain.wallet import Wallet

    wallet = Wallet('private_key_string')
    client = HttpApiClient()

    multi_transfer_msg = TransferMsg(
        wallet=wallet,
        transfers=[
            Transfer(symbol='ETH.B', amount=1),
            Transfer(symbol='BNB', amount=1),
        ],
        to_address='<to address>',
        memo='Thanks for the beer'
    )
    res = client.broadcast_msg(multi_transfer_msg, sync=True)

**Vote for proposal**

.. code:: python

    from binance_chain.http import HttpApiClient
    from binance_chain.messages import VoteMsg

docs/changelog.rst
Changelog
=========

v0.1.20 - 2019-06-29
^^^^^^^^^^^^^^^^^^^^

**Added**

- Multi Transfer Msg option

**Fixed**

- Response code of 0 for http requests

v0.1.19 - 2019-04-30
^^^^^^^^^^^^^^^^^^^^

**Added**

- Shuffling of peer nodes in the Pooled HTTP RPC Node client

v0.1.18 - 2019-04-29
^^^^^^^^^^^^^^^^^^^^

**Added**

- Advanced async Pooled HTTP RPC Node client spreading requests over available peers

**Updated**

- RPC request id behaviour to work across multiple connections

v0.1.17 - 2019-04-29
^^^^^^^^^^^^^^^^^^^^

**Added**

- UltraJson ultra fast JSON parser library
- Docs for requests/aiohttp params and proxy support
- more docs and tests

**Fixed**

- DepthCache close method

v0.1.16 - 2019-04-28
^^^^^^^^^^^^^^^^^^^^

**Added**

- Vote transaction type
- Simple transaction signing example

**Fixed**

- Depth Cache doesn't wait for websocket messages before outputting

v0.1.13 - 2019-04-28
^^^^^^^^^^^^^^^^^^^^

**Added**

- `get_address` function for Ledger hardware wallet
- better error handling and parsing of Ledger hardware wallet responses

v0.1.12 - 2019-04-27
^^^^^^^^^^^^^^^^^^^^

**Added**

- Ledger Hardware wallet support for signing transactions

v0.1.10 - 2019-04-24
^^^^^^^^^^^^^^^^^^^^

**Fixed**

- Connecting to secure and insecure websocket connections

v0.1.9 - 2019-04-23
^^^^^^^^^^^^^^^^^^^

**Added**

- More test coverage including for python 3.7

**Fixed**

- Params in async http client for `get_klines`
- coveralls report
- small fixes

v0.1.8 - 2019-04-21
^^^^^^^^^^^^^^^^^^^

**Added**

- `get_block_exchange_fee` function to Http API Clients
- memo param to Transfer message
- more tests

**Updated**

- Remove jsonrpcclient dependency

**Fixed**

- Use of enums in request params
- Some deprecation warnings

v0.1.7 - 2019-04-18
^^^^^^^^^^^^^^^^^^^

**Updated**

- fix package


v0.1.6 - 2019-04-17
^^^^^^^^^^^^^^^^^^^

**Updated**

- fix package requirement versions

v0.1.5 - 2019-04-17
^^^^^^^^^^^^^^^^^^^

**Fixed**

- signing module imported

v0.1.4 - 2019-04-16
^^^^^^^^^^^^^^^^^^^

**Fixed**

- Issue with protobuf file

v0.1.3 - 2019-04-16
^^^^^^^^^^^^^^^^^^^

**Added**

- Wallet methods for Binance Signing Service v0.0.2

v0.1.2 - 2019-04-14
^^^^^^^^^^^^^^^^^^^

**Added**

- Binance Chain Signing Service Interfaces v0.0.1

**Updated**

- Cleaned up TransferMsg as from_address is found from wallet instance

v0.1.1 - 2019-04-13
^^^^^^^^^^^^^^^^^^^

**Added**

- Broadcast message taking signed hex data

v0.1.0 - 2019-04-11
^^^^^^^^^^^^^^^^^^^

**Added**

- Async versions of HTTP Client
- Async version of Node RPC Client
- Node RPC Websocket client
- Async Depth Cache
- Transfer message implementation
- Message broadcast over Node RPC

v0.0.5 - 2019-04-08
^^^^^^^^^^^^^^^^^^^

**Added**

- All websocket stream endpoints
- Wallet functions to read account and keep track of transaction sequence
- Support for Testnet and Production environments, along with user defined environment
- Helper classes to create limit buy and sell messages

**Updated**

- Refactored modules and tidied up message creation and wallets

v0.0.4 - 2019-04-07
^^^^^^^^^^^^^^^^^^^

**Added**

- Wallet initialise from private key or mnemonic string
- Create wallet by generating a mnemonic

v0.0.3 - 2019-04-06
^^^^^^^^^^^^^^^^^^^

**Added**

- Transaction Broadcasts
- Generated Docs

v0.0.2 - 2019-04-04
^^^^^^^^^^^^^^^^^^^

**Added**

- NodeRPC implementation
- Websockets

v0.0.1 - 2019-02-24
^^^^^^^^^^^^^^^^^^^

- HTTP API Implementation


binance_chain/client.py
import logging
from typing import Optional, Dict

import requests

import binance_chain.messages
from binance_chain.environment import BinanceEnvironment
from binance_chain.constants import PeerType, KlineInterval, OrderSide, OrderStatus, TransactionSide, TransactionType
from binance_chain.exceptions import (
    BinanceChainAPIException, BinanceChainRequestException, BinanceChainBroadcastException
)


class HttpApiClient:

    API_VERSION = 'v1'

    def __init__(self, env: Optional[BinanceEnvironment] = None, requests_params: Optional[Dict] = None):
        """Binance Chain API Client constructor

        https://binance-chain.github.io/api-reference/dex-api/paths.html

        :param env: (optional) A BinanceEnvironment instance or None which will default to production env
        :param requests_params: (optional) Dictionary of requests params to use for all calls
        :type requests_params: dict.

        .. code:: python

            # get production env client
            prod_client = Client()

            # get testnet env client
            testnet_env = BinanceEnvironment.get_testnet_env()
            client = Client(testnet_env)

        """

        self._env = env or BinanceEnvironment.get_production_env()
        self._requests_params = requests_params
        self.session = self._init_session()

    @property
    def env(self):
        return self._env

    def _init_session(self):

        session = requests.session()
        headers = {
            'Accept': 'application/json',
            'User-Agent': 'python-binance-chain',
        }
        session.headers.update(headers)
        return session

    def _create_path(self, path):
        return '/api/{}/{}'.format(self.API_VERSION, path)

    def _create_uri(self, path):
        return '{}{}'.format(self._env.api_url, path)

    def _request(self, method, path, **kwargs):

        # set default requests timeout
        kwargs['timeout'] = 10

        # add our global requests params
        if self._requests_params:
            kwargs.update(self._requests_params)

        kwargs['data'] = kwargs.get('data', {})
        kwargs['headers'] = kwargs.get('headers', {})

        full_path = self._create_path(path)
        uri = self._create_uri(full_path)
        print(f"uri:{uri}")

        if kwargs['data'] and method == 'get':
            kwargs['params'] = kwargs['data']
            del(kwargs['data'])

        if method == 'post':
            kwargs['headers']['content-type'] = 'text/plain'

        response = getattr(self.session, method)(uri, **kwargs)
        return self._handle_response(response)

    @staticmethod
    def _handle_response(response):
        """Internal helper for handling API responses from the server.
        Raises the appropriate exceptions when necessary; otherwise, returns the
        response.
        """
        if not str(response.status_code).startswith('2'):
            raise BinanceChainAPIException(response)
        try:
            res = response.json()

            if 'code' in res and res['code'] != "200000":
                raise BinanceChainAPIException(response)

            if 'success' in res and not res['success']:
                raise BinanceChainAPIException(response)

            # by default return full response
            # if it's a normal response we have a data attribute, return that
            if 'data' in res:
                res = res['data']
            return res
        except ValueError:
            raise BinanceChainRequestException('Invalid Response: %s' % response.text)

    def _get(self, path, **kwargs):
        return self._request('get', path, **kwargs)

    def _post(self, path, **kwargs):
        return self._request('post', path, **kwargs)

    def _put(self, path, **kwargs):
        return self._request('put', path, **kwargs)

    def _delete(self, path, **kwargs):
        return self._request('delete', path, **kwargs)

    def get_time(self):
        """Get the server timestamp

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1time

        .. code:: python

            time = client.get_time()

        :return: API Response

        .. code-block:: python

            {
                "ap_time": "2019-02-24T09:23:35Z",
                "block_time": "2019-02-24T09:23:34Z"
            }

        """
        return self._get("time")

    def get_node_info(self):
        """Get node runtime information

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1node-info

        .. code:: python

            node_info = client.get_node_info()

        :return: API Response

        """
        return self._get("node-info")

    def get_validators(self):
        """Gets the list of validators used in consensus

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1validators

        .. code:: python

            validators = client.get_validators()

        :return: API Response

        """
        return self._get("validators")

    def get_peers(self, peer_type: Optional[PeerType] = None):
        """Gets the list of network peers

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1peers

        .. code:: python

            peers = client.get_peers()

        :return: API Response

        """
        peers = self._get("peers")
        if peer_type:
            peers = [p for p in peers if peer_type in p['capabilities']]

        return peers

    def get_node_peers(self):
        """Get a list of peers with Node RPC support

        :return:
        """
        return self.get_peers(peer_type=PeerType.NODE)

    def get_websocket_peers(self):
        """Get a list of peers with Websocket support

        :return:
        """
        return self.get_peers(peer_type=PeerType.WEBSOCKET)

    def get_account(self, address: str):
        """Gets account metadata for an address

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1accountaddress

        .. code:: python

            account = client.get_account('tbnb185tqzq3j6y7yep85lncaz9qeectjxqe5054cgn')

        :return: API Response

        """
        return self._get(f"account/{address}")

    def get_account_sequence(self, address: str):
        """Gets an account sequence for an address.

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1accountaddresssequence

        .. code:: python

            account_seq = client.get_account_sequence('tbnb185tqzq3j6y7yep85lncaz9qeectjxqe5054cgn')

        :return: API Response

        .. code-block:: python

        """
        return self._get(f"account/{address}/sequence")

    def get_transaction(self, transaction_hash: str):
        """Gets transaction metadata by transaction ID

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1txhash

        .. code:: python

            account_seq = client.get_transaction('E81BAB8E555819E4211D62E2E536B6D5812D3D91C105F998F5C6EB3AB8136482')

        :return: API Response

        """

        data = {
            'format': 'json'
        }

        return self._get(f"tx/{transaction_hash}?format=json", data=data)

    def get_tokens(self):
        """Gets a list of tokens that have been issued

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1tokens

        .. code:: python

            tokens = client.get_tokens()

        :return: API Response

        """
        return self._get("tokens")

    def get_markets(self):
        """Gets the list of market pairs that have been listed

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1markets

        .. code:: python

            markets = client.get_markets()

        :return: API Response

        """
        return self._get("markets")

    def get_fees(self):
        """Gets the current trading fees settings

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1fees

        .. code:: python

            fees = client.get_fees()

        :return: API Response

        """
        return self._get("fees")

    def get_order_book(self, symbol: str):
        """Gets the order book depth data for a given pair symbol

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1depth

        :param symbol: required e.g NNB-0AD_BNB

        .. code:: python

            order_book = client.get_order_book('NNB-0AD_BNB')

        :return: API Response

        """
        data = {
            'symbol': symbol
        }
        return self._get("depth", data=data)

    def broadcast_msg(self, msg: binance_chain.messages.Msg, sync: bool = False):
        """Broadcast a message

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1broadcast

        :param msg: Type of NewOrderMsg, CancelOrderMsg, FreezeMsg, UnfreezeMsg
        :param sync: Synchronous broadcast (wait for DeliverTx)?

        .. code:: python

            # new order example
            # construct the message
            new_order_msg = NewOrderMsg(
                symbol="ANN-457_BNB",
                time_in_force=TimeInForce.GTE,
                order_type=OrderType.LIMIT,
                side=OrderSide.BUY,
                price=Decimal(0.000396000),
                quantity=Decimal(12)
            )
            # then broadcast it
            res = client.broadcast_msg(new_order_msg, wallet)

            # cancel order example
            cancel_order_msg = CancelOrderMsg(
                order_id="09F8B32D33CBE2B546088620CBEBC1FF80F9BE001ACF42762B0BBFF0A729CE3",
                symbol='ANN-457_BNB',
            )
            res = client.broadcast_msg(cancel_order_msg, wallet)

        :return: API Response

        """

        # fetch account detail
        # account = self.get_account(self.msg.wallet.address)

        if self._env != msg.wallet.env:
            raise BinanceChainBroadcastException("Wallet environment doesn't match HttpApiClient environment")

        msg.wallet.initialise_wallet(self)
        data = msg.to_hex_data()

        logging.debug(f'data:{data}')

        req_path = 'broadcast'
        if sync:
            req_path += f'?sync=1'

        res = self._post(req_path, data=data)
        msg.wallet.increment_account_sequence()
        return res

    def get_klines(self, symbol: str, interval: KlineInterval, limit: Optional[int] = 300,
                   start_time: Optional[int] = None, end_time: Optional[int] = None):
        """Gets candlestick/kline bars for a symbol. Bars are uniquely identified by their open time

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1klines

        :param symbol: required e.g NNB-0AD_BNB
        :param interval: required e.g 1m, 3m, 5m, 15m, 30m, 1h, 2h, 4h, 6h, 8h, 12h, 1d, 3d, 1w, 1M
        :param limit:
        :param start_time:
        :param end_time:

        .. code:: python

            klines = client.get_klines('NNB-0AD_BNB', KlineInterval.ONE_DAY)

        :return: API Response

        """
        data = {
            'symbol': symbol,
            'interval': interval
        }
        if limit is not None:
            data['limit'] = limit
        if start_time is not None:
            data['startTime'] = start_time
        if end_time is not None:
            data['endTime'] = end_time

        return self._get("klines", data=data)

    def get_closed_orders(
        self, address: str, symbol: Optional[str] = None, status: Optional[OrderStatus] = None,
        side: Optional[OrderSide] = None, offset: Optional[int] = 0, limit: Optional[int] = 500,
        start_time: Optional[int] = None, end_time: Optional[int] = None, total: Optional[int] = 0
    ):
        """Gets closed (filled and cancelled) orders for a given address

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1ordersclosed

        :param address: required
        :param symbol: e.g NNB-0AD_BNB
        :param status: order status type
        :param side: order side. 1 for buy and 2 for sell.
        :param offset: start with 0; default 0.
        :param limit: default 500; max 1000.
        :param start_time:
        :param end_time:
        :param total: total number required, 0 for not required and 1 for required, default not required, return total=-1

        .. code:: python

            orders = client.get_closed_orders('tbnb185tqzq3j6y7yep85lncaz9qeectjxqe5054cgn')

        :return: API Response

        """
        data = {
            'address': address
        }
        if symbol is not None:
            data['symbol'] = symbol
        if status is not None:
            data['status'] = status
        if side is not None:
            data['side'] = side
        if offset is not None:
            data['offset'] = offset
        if limit is not None:
            data['limit'] = limit
        if start_time is not None:
            data['start'] = start_time
        if end_time is not None:
            data['end'] = end_time
        if total is not None:
            data['total'] = total

        return self._get("orders/closed", data=data)

    def get_open_orders(
        self, address: str, symbol: Optional[str] = None, offset: Optional[int] = 0, limit: Optional[int] = 500,
        total: Optional[int] = 0
    ):
        """Gets open orders for a given address

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1ordersopen

        :param address: required
        :param symbol: e.g NNB-0AD_BNB
        :param offset: start with 0; default 0.
        :param limit: default 500; max 1000.
        :param total: total number required, 0 for not required and 1 for required, default not required, return total=-1

        .. code:: python

            orders = client.get_open_orders('tbnb185tqzq3j6y7yep85lncaz9qeectjxqe5054cgn')

        :return: API Response

        """
        data = {
            'address': address
        }
        if symbol is not None:
            data['symbol'] = symbol
        if offset is not None:
            data['offset'] = offset
        if limit is not None:
            data['limit'] = limit
        if total is not None:
            data['total'] = total

        return self._get("orders/open", data=data)

    def get_order(self, order_id: str):
        """Gets metadata for an individual order by its ID

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1ordersid

        :param order_id: required

        .. code:: python

            orders = client.get_order('')

        :return: API Response

        """

        return self._get(f"orders/{order_id}")

    def get_ticker(self, symbol: str):
        """Gets 24 hour price change statistics for a market pair symbol

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1ticker24hr

        :param symbol: required

        .. code:: python

            ticker = client.get_ticker('NNB-0AD_BNB')

        :return: API Response

        """
        data = {
            'symbol': symbol
        }

        return self._get("ticker/24hr", data=data)

    def get_trades(
        self, address: Optional[str] = None, symbol: Optional[str] = None,
        side: Optional[OrderSide] = None, quote_asset: Optional[str] = None, buyer_order_id: Optional[str] = None,
        seller_order_id: Optional[str] = None, height: Optional[str] = None, offset: Optional[int] = 0,
        limit: Optional[int] = 500, start_time: Optional[int] = None, end_time: Optional[int] = None,
        total: Optional[int] = 0
    ):
        """Gets a list of historical trades

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1trades

        :param address:
        :param symbol:
        :param side:
        :param quote_asset:
        :param buyer_order_id:
        :param seller_order_id:
        :param height:
        :param offset:
        :param limit:
        :param start_time:
        :param end_time:
        :param total:

        .. code:: python

            trades = client.get_trades('')

        :return: API Response

        """
        data = {}
        if address is not None:
            data['address'] = address
        if symbol is not None:
            data['symbol'] = symbol
        if side is not None:
            data['side'] = side
        if quote_asset is not None:
            data['quoteAsset'] = quote_asset
        if buyer_order_id is not None:
            data['buyerOrderId'] = buyer_order_id
        if seller_order_id is not None:
            data['sellerOrderId'] = seller_order_id
        if height is not None:
            data['height'] = height
        if offset is not None:
            data['offset'] = offset
        if limit is not None:
            data['limit'] = limit
        if start_time is not None:
            data['start'] = start_time
        if end_time is not None:
            data['end'] = end_time
        if total is not None:
            data['total'] = total

        return self._get("trades", data=data)

    def get_transactions(
        self, address: str, symbol: Optional[str] = None,
        side: Optional[TransactionSide] = None, tx_asset: Optional[str] = None,
        tx_type: Optional[TransactionType] = None, height: Optional[str] = None, offset: Optional[int] = 0,
        limit: Optional[int] = 500, start_time: Optional[int] = None, end_time: Optional[int] = None
    ):
        """Gets a list of transactions

        https://binance-chain.github.io/api-reference/dex-api/paths.html#apiv1transactions

        :param address:
        :param symbol:
        :param side:
        :param tx_asset:
        :param tx_type:
        :param height:
        :param offset:
        :param limit:
        :param start_time:
        :param end_time:

        .. code:: python

            transactions = client.get_transactions('')

        :return: API Response

        """
        data = {
            'address': address
        }
        if symbol is not None:
            data['symbol'] = symbol
        if side is not None:
            data['side'] = side
        if tx_asset is not None:
            data['txAsset'] = tx_asset
        if tx_type is not None:
            data['txType'] = tx_type
        if height is not None:
            data['blockHeight'] = height
        if offset is not None:
            data['offset'] = offset
        if limit is not None:
            data['limit'] = limit
        if start_time is not None:
            data['start'] = start_time
        if end_time is not None:
            data['end'] = end_time

        return self._get("transactions", data=data)
